src/os/bluestore/BlueFS.cc

```diff
@@ -2269,6 +2426,185 @@ int64_t BlueFS::_read_random(
+    File::WALFlush new_flush(flush_offset, flush_length);
+
+    // read marker
+    bl.clear();
+    uint64_t marker_offset = new_flush.get_marker_offset();
+    read_result = _read(h, marker_offset, new_flush.tail_size(), &bl, nullptr);
+    if (read_result < new_flush.tail_size()) {
+      dout(20) << fmt::format("{} cannot read marker, most likely we are out of bounds. flush_offset={:#X}, marker_offset={:#X}", __func__, flush_offset, marker_offset) << dendl;
+      break;
+    }
+    uint64_t marker;
+    buffer_iterator = bl.cbegin();
+    decode(marker, buffer_iterator);
+    if (marker != File::WALFlush::generate_hashed_marker(super.osd_uuid, file->fnode.ino)) {
+      // EOF or corruption
+      dout(30) << fmt::format("reached eof or marker corruption {:#x}", flush_offset) << dendl;
+      break;
+    }
+
+    uint64_t increase = new_flush.end_offset() - new_flush.offset;
+    dout(20) << fmt::format("{} adding flush {:#x}~{:#x}", __func__, flush_offset, new_flush.length) << dendl;
+    file->wal_flushes.push_back(new_flush);
+    if (flush_offset >= flush_end) {
+      dout(20) << fmt::format("{} recovering flush {:#x}~{:#x}", __func__, flush_offset, new_flush.length) << dendl;
+      file->fnode.wal_size += new_flush.length;
+      file->fnode.size += increase;
+      vselector->add_usage(file->vselector_hint, increase);
+    }
+
+    flush_offset += increase;
+  }
+
+  // if we read less it might mean corruption
+  if (flush_offset < flush_end) {
+    dout(20) << fmt::format("{} read less than expected {:#x} bytes", __func__, flush_offset) << dendl;
+  }
+  ceph_assert(flush_offset >= flush_end);
+
+  delete h;
+}
+
+int64_t BlueFS::_read_wal(
+  FileReader *h,         ///< [in] read from here
+  uint64_t off,          ///< [in] offset
+  size_t len,            ///< [in] this many bytes
+  bufferlist *outbl,     ///< [out] optional: reference the result here
+  char *out)             ///< [out] optional: or copy it here
+{
+  ceph_assert(h->file->is_wal_read_loaded);
+  dout(20) << __func__ << " h " << h << " offset: 0x" 
+    << off << std::hex << "~" << len << std::hex << dendl;
+  if (outbl) {
+    outbl->clear();
+  }
+
+  int64_t ret = 0;
+
+  // WAL data is wrapped in an envelope that has a format of [length of flush, payload, file ino]
+  // wal_data_logical_offset points to the offset of the payload we are currently in.
+  uint64_t wal_data_logical_offset = 0;
+
+
+  // save previous position as buffer pos is treated difffernt on regular files
+  uint64_t previous_pos = h->buf.pos;
+
+  uint64_t remaining_len = len;
+  auto flush_iterator = h->file->wal_flushes.begin();
+  while (remaining_len > 0 && flush_iterator != h->file->wal_flushes.end()) {
+    uint64_t flush_offset = flush_iterator->offset;
+    uint64_t flush_length = flush_iterator->length;
+    dout(25) << fmt::format("{} flush_offset={:#x} flush_length={:#x}", __func__, flush_offset, flush_length) << dendl;
+
+    if (flush_length == 0) {
```

Comment:
Is it even legal / possible to have FLUSH of size 0?
We should create new WALFlush frame on first append after flush to disk, so there should be somehting.